sdk/python/feast/stream_feature_view.py

```diff
@@ -1,32 +1,112 @@
+import abc
+import functools
+import warnings
 from datetime import timedelta
+from types import MethodType
 from typing import Dict, List, Optional, Union
 
-from feast.data_source import DataSource
+import dill
+from google.protobuf.duration_pb2 import Duration
+
+from feast.data_source import DataSource, KafkaSource
 from feast.entity import Entity
 from feast.feature_view import FeatureView
 from feast.field import Field
 from feast.protos.feast.core.DataSource_pb2 import DataSource as DataSourceProto
+from feast.protos.feast.core.OnDemandFeatureView_pb2 import (
+    UserDefinedFunction as UserDefinedFunctionProto,
+)
+from feast.protos.feast.core.StreamFeatureView_pb2 import (
+    Aggregation as AggregationProto,
+)
+from feast.protos.feast.core.StreamFeatureView_pb2 import (
+    StreamFeatureView as StreamFeatureViewProto,
+)
+from feast.protos.feast.core.StreamFeatureView_pb2 import (
+    StreamFeatureViewMeta as StreamFeatureViewMetaProto,
+)
+from feast.protos.feast.core.StreamFeatureView_pb2 import (
+    StreamFeatureViewSpec as StreamFeatureViewSpecProto,
+)
+
+warnings.simplefilter("once", RuntimeWarning)
+
+SUPPORTED_STREAM_SOURCES = {"KafkaSource", "PushSource"}
+
+
+class Aggregation(abc.ABC):
+    """
+    NOTE: Feast-handled aggregations are not yet supported. This class provides a way to register user-defined aggregations.
+    """
+
+    column: str  # Column name of the feature we are aggregating.
+    function: str  # Provided built in aggregations sum, max, min, count mean
+    time_windows: List[str]  # The time window. Example ["1h", "24h"]
+
+    def __init__(self, column: str, function: str, time_windows: List[str]):
+        self.column = column
+        self.function = function
+        self.time_windows = time_windows
+
+    def to_proto(self) -> AggregationProto:
+        return AggregationProto(
+            column=self.column, function=self.function, time_windows=self.time_windows,
+        )
+
+    @classmethod
+    def from_proto(cls, agg_proto: AggregationProto):
+        aggregation = cls(
+            column=agg_proto.column,
+            function=agg_proto.function,
+            time_windows=list(agg_proto.time_windows),
+        )
+        return aggregation
+
+    def __eq__(self, other):
+        if not isinstance(other, Aggregation):
+            raise TypeError("Comparisons should only involve Aggregations.")
+
+        if (
+            self.column != other.column
+            or self.function != other.function
+            or self.time_windows != other.time_windows
```

Comment:
might need to sort this list? depends on our exact assumptions, just wanted to call it out

Reply:
good catch!